### DISCUSSION TITLE
Question about async handling

### DISCUSSION THREAD

**Discussion by junyilou** [python-telegram-bot/python-telegram-bot#2904] [Q&A] [answered]
Hello guys, I have a question that has been bothering me for a long time. I would greatly appreciate your help!

Let's imagine a situation like this: 

> You entered a function by a command. In this particular function, there is a loop that takes a long time to finish, or worse, it is actually an infinite loop. The question is, how can you abort this loop or function so that the bot can handle other requests?

An example would be like this:

```python
def function(update, context):
  counter = 0
  message = update.message.reply_text(str(counter))

  while True:
    counter += 1
    message = message.edit_text(text = str(counter))
    time.sleep(1)

dispatcher.add_handler(CommandHandler("main", function))
```

After user enters this function by "/main", he or she gets a message that includes a number, which the number increases every second and this will take forever to finish.

Something that I have tried:

- Design another `MessageHandler`, if user sends some keywords like "stop", it changes the value of a global variable. In the loop, it checks if this variable is `True` every time before it adds the number, if it was `False`, it breaks. 

  However, the MessageHandler won't detect this message before the loop has actually ended, so this is meaningless.

- Include a `InlineKeyboard` in the message, with a button called "Stop". Design another `CallbackQueryHandler` to handle this button. 

  Similarly, CallbackQueryHandler will only receive the callback after the loop was finished. This can be verified if I set a manual `break` when the counter hits a number like 5 or something.

I was thinking if I could design some kind of abort function that runs asynchronously, but I am personally not very familiar with async in Python. After some trials like setting `async def` or `CallbackQueryHandler(query_handle, run_async = True)`, I still failed to produce this effect.

I noticed there is a decorator called `from telegram.ext.dispatcher import run_async`, does that help? Or, is there any other way to realize this result?

Thanks!

----

Note: This question was also asked on StackOverflow: https://stackoverflow.com/questions/71157034

**Comment by Happy-Ferret** [comment, marked as answer]
If you do things asynchronously, it won't even matter if it aborts or not because it's non-blocking.

Personally, I use an `aexec` decorator (I'd like to get rid of that design, alas) to achieve async.

Your code would look like this:

```python
import asyncio

def aexec(func):
    def wrapper(update: Update, context: CallbackContext):
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        loop.run_until_complete(func(update, context))
        loop.close()
    return wrapper

@aexec
async def function(update, context):
  counter = 0
  message = update.message.reply_text(str(counter))

  while True:
    counter += 1
    message = message.edit_text(text = str(counter))
    await asyncio.sleep(1)

def main():
...
dispatcher.add_handler(CommandHandler("main", function, run_async=True))
```

Note the `run_async=True` in the `add_handler` call. This is important.

**Reply by junyilou** [reply]
Thank you, I have utilized your decorator and managed to pull this off!

**Reply by Happy-Ferret** [reply]
> Thank you, I have utilized your decorator and managed to pull this off!

Glad to know it works for you and happy to help.

I'm working on my own library that wraps parts of PTB.
That decorator is something I mostly use internally, as I've now added an `@async_command(name="name_of_command")` decorator to my library that removes the need for `add_handler` while also automatically calling `@aexec`.
